packages/rum/src/domain/record/record.spec.ts

```diff
@@ -196,6 +213,191 @@ describe('record', () => {
+        IncrementalSource.Mutation
+      )
+      expect(innerMutationData.adds.length).toBe(1)
+      expect(innerMutationData.adds[0].node.type).toBe(2)
+      expect(innerMutationData.adds[0].parentId).toBe(host.id)
+      const addedNode = innerMutationData.adds[0].node as ElementNode
+      expect(addedNode.tagName).toBe('span')
+    })
+
+    it('should record mutation inside a shadow root added after the FS', () => {
+      updateExperimentalFeatures(['record_shadow_dom'])
+      startRecording()
+      expect(getEmittedRecords().length).toBe(recordsPerFullSnapshot())
+
+      // shadow DOM mutation
+      const span = document.createElement('span')
+      span.className = 'toto'
+      createShadow([span])
+      recordApi.flushMutations()
+      expect(getEmittedRecords().length).toBe(recordsPerFullSnapshot() + 1)
+      const hostMutationData = getLastIncrementalSnapshotData<BrowserMutationData>(
+        getEmittedRecords(),
+        IncrementalSource.Mutation
+      )
+      expect(hostMutationData.adds.length).toBe(1)
+      const hostNode = hostMutationData.adds[0].node as ElementNode
+      expect(hostNode.isShadowHost).toBe(true)
+
+      // inner mutation
+      span.className = 'titi'
+      recordApi.flushMutations()
+      expect(getEmittedRecords().length).toBe(recordsPerFullSnapshot() + 2)
+      const innerMutationData = getLastIncrementalSnapshotData<BrowserMutationData>(
+        getEmittedRecords(),
+        IncrementalSource.Mutation
+      )
+      expect(innerMutationData.attributes.length).toBe(1)
+      expect(innerMutationData.attributes[0].attributes.class).toBe('titi')
+    })
+
+    it('should record the change event inside a shadow root', () => {
+      updateExperimentalFeatures(['record_shadow_dom'])
+      const radio = document.createElement('input')
+      radio.setAttribute('type', 'radio')
+      createShadow([radio])
+      startRecording()
+      expect(getEmittedRecords().length).toBe(recordsPerFullSnapshot())
+
+      // inner mutation
+      radio.checked = true
+      radio.dispatchEvent(createNewEvent('change', { target: radio, composed: false }))
+
+      recordApi.flushMutations()
+      const innerMutationData = getLastIncrementalSnapshotData<BrowserMutationData & { isChecked: boolean }>(
+        getEmittedRecords(),
+        IncrementalSource.Input
+      )
+      expect(innerMutationData.isChecked).toBe(true)
+    })
+
+    it('should clean the state once the shadow dom is removed to avoid memory leak', () => {
+      updateExperimentalFeatures(['record_shadow_dom'])
+      const div = document.createElement('div')
+      div.className = 'toto'
+      const shadowRoot = createShadow([div])
+      startRecording()
+      expect(getEmittedRecords().length).toBe(recordsPerFullSnapshot())
+      expect(recordApi.shadowDomCallBacks.size).toBe(1)
+      shadowRoot.host.remove()
+      recordApi.flushMutations()
+      expect(recordApi.shadowDomCallBacks.size).toBe(0)
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏IMO having unit test for the shadow dom file to test the behavior of adding/removing shadow roots could be enough and would avoid to expose the map through recordingAPI.

wdyt?

Reply:
It was an important thing that I missed and would create memory leaks therefore I added a unit tests to avoid introducing a regression that being said I agree that we're testing an implementation details. I can delete it if you want